src/ast/nodes/shared/ClassNode.ts

```diff
@@ -1,47 +1,242 @@
+import { getOrCreate } from '../../../utils/getOrCreate';
 import { CallOptions } from '../../CallOptions';
+import { DeoptimizableEntity } from '../../DeoptimizableEntity';
 import { HasEffectsContext } from '../../ExecutionContext';
 import ChildScope from '../../scopes/ChildScope';
 import Scope from '../../scopes/Scope';
-import { ObjectPath } from '../../utils/PathTracker';
+import { EMPTY_PATH, ObjectPath, PathTracker } from '../../utils/PathTracker';
+import { LiteralValueOrUnknown, UnknownValue, UNKNOWN_EXPRESSION } from '../../values';
 import ClassBody from '../ClassBody';
 import Identifier from '../Identifier';
+import MethodDefinition from '../MethodDefinition';
+import { ExpressionEntity } from './Expression';
 import { ExpressionNode, NodeBase } from './Node';
 
 export default class ClassNode extends NodeBase {
 	body!: ClassBody;
 	id!: Identifier | null;
 	superClass!: ExpressionNode | null;
+	private classConstructor!: MethodDefinition | null;
+	private deoptimizedPrototype = false;
+	private deoptimizedStatic = false;
+	// Collect deoptimization information if we can resolve a property access, by property name
+	private expressionsToBeDeoptimized = new Map<string, DeoptimizableEntity[]>();
+	// Known, simple, non-deoptimized static properties are kept in here. They are removed when deoptimized.
+	private staticPropertyMap: {[name: string]: ExpressionNode} | null = null;
+
+	bind() {
+		super.bind();
+	}
 
 	createScope(parentScope: Scope) {
 		this.scope = new ChildScope(parentScope);
 	}
 
-	hasEffectsWhenAccessedAtPath(path: ObjectPath) {
-		if (path.length <= 1) return false;
-		return path.length > 2 || path[0] !== 'prototype';
+	deoptimizeAllStatics() {
+		for (const name in this.staticPropertyMap) {
+			this.deoptimizeStatic(name);
+		}
+		this.deoptimizedStatic = this.deoptimizedPrototype = true;
+	}
+
+	deoptimizeCache() {
+		this.deoptimizeAllStatics();
```

Comment:
I do not believe this is necessary unless you rely on retrieving literal values for computed properties. That is also why you cannot test it.

So it works like this:
When you perform a `getLiteralValueAtPath` or a `getReturnExpressionWhenCalledAtPath` you pass a `DeoptimizableEntity` as the last parameter (which is usually the Node requesting the literal/return value). The reason is that it is possible a literal/return value might be deoptimized at a later time when binding/tree-shaking progresses. If that happens, then some Node down the chain that is responsible for the deoptimization will call `deoptimizeCache` on this entity. At this point in time, the entity is responsible for "forgetting" the cached value and also deoptimize the cache of all dependent entities that may rely on that value.

Admittedly it is kind of difficult to describe such deoptimization scenarios. Ideally I would just put a breakpoint into "deoptimizeCache" of ObjectExpression and see what test triggers it.

This will become even more important in the future as one goal is to move more and more of the initial deoptimizations to "on demand late deoptimization" i.e. only deoptimize when as reassignment is included.

Reply:
> I do not believe this is necessary unless you rely on retrieving literal values for computed properties.

You mean this is only necessary for nodes that themselves retrieve literal values? Do I understand correctly that the call-throughs to definition nodes in the class' own `getLiteralValueAtPath` and `getReturnExpressionWhenCalledAtPath` don't count because they just pass through their argument instead of passing themselves?